src/edge/acrManager.ts

```diff
@@ -0,0 +1,224 @@
+// Licensed under the MIT license.
+
+"use strict";
+import { ContainerRegistryManagementClient } from "azure-arm-containerregistry";
+import { Registry, RegistryListCredentialsResult } from "azure-arm-containerregistry/lib/models";
+import * as request from "request-promise";
+import * as vscode from "vscode";
+import { Executor } from "../common/executor";
+import { UserCancelledError } from "../common/UserCancelledError";
+import { Utility } from "../common/utility";
+import { AcrRegistryQuickPickItem } from "../container/models/AcrRegistryQuickPickItem";
+import { AzureAccount, AzureSession, AzureSubscription } from "../typings/azure-account.api";
+
+export class AcrManager {
+    private readonly azureAccount: AzureAccount;
+    private refreshTokenArc: string;
+
+    constructor() {
+        this.azureAccount = vscode.extensions.getExtension<AzureAccount>("ms-vscode.azure-account")!.exports;
+    }
+
+    public async selectAcrImage(): Promise<string> {
+        const acrRegistryItem: AcrRegistryQuickPickItem = await this.selectAcrRegistry();
+        if (acrRegistryItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        const registryUrl: string = acrRegistryItem.registry.loginServer;
+
+        const acrRepoItem: vscode.QuickPickItem = await this.selectAcrRepo(registryUrl, acrRegistryItem.azureSubscription.session);
+        if (acrRepoItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        const acrTagItem: vscode.QuickPickItem = await this.selectAcrTag(registryUrl, acrRepoItem.label);
+        if (acrTagItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        this.loginToAcr(acrRegistryItem.registry, acrRegistryItem.azureSubscription);
+
+        return acrTagItem.description;
+    }
+
+    private async selectAcrRegistry(): Promise<AcrRegistryQuickPickItem> {
+        if (!(await this.azureAccount.waitForLogin())) {
+            await vscode.commands.executeCommand("azure-account.askForLogin");
+        }
+
+        const acrRegistryItem: AcrRegistryQuickPickItem = await vscode.window.showQuickPick(this.loadAcrRegistryItems(), { placeHolder: "Select Azure Container Registry", ignoreFocusOut: true });
+        return acrRegistryItem;
+    }
+
+    private async loadAcrRegistryItems(): Promise<AcrRegistryQuickPickItem[]> {
+        try {
+            await this.azureAccount.waitForFilters();
+            const registryPromises: Array<Promise<AcrRegistryQuickPickItem[]>> = [];
+            for (const azureSubscription of this.azureAccount.filters) {
+                const client = new ContainerRegistryManagementClient(
+                    azureSubscription.session.credentials,
+                    azureSubscription.subscription.subscriptionId!,
+                );
+
+                registryPromises.push(
+                    Utility.listAllAzureResource(client.registries, client.registries.list())
```

Comment:
use await Utility.listAllAzureResource, instead of then? to make code more easily read

Reply:
Adding `await` here will suspend the execution and thus affects the performance.